File: apps/backend/api/proactive_sms.py
"""
Proactive SMS trigger endpoint.

Legacy authenticated delivery surface. Internal scheduling is owned by FastAPI;
any stale external call enters the same durable occurrence fence.

Auth: requires x-internal-secret header matching INTERNAL_SMS_CHAT_SECRET.

Usage:
  - POST /api/sms/proactive/trigger  {"trigger_type": "eod_recap"}
      → runs a single trigger type at its default hour
  - POST /api/sms/proactive/trigger  {"trigger_type": "all"}
      → runs ALL trigger types, each at its default hour
      → designed to be called once per hour by a cron
"""

import logging
import os
from typing import Optional

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel

logger = logging.getLogger(__name__)

INTERNAL_SMS_CHAT_SECRET = os.getenv("INTERNAL_SMS_CHAT_SECRET", "")


class ProactiveTriggerRequest(BaseModel):
    trigger_type: str = "eod_recap"
    target_hour: Optional[int] = None  # None = use default for trigger type


router = APIRouter()


@router.post("/api/sms/proactive/trigger")
async def proactive_trigger(request: Request, body: ProactiveTriggerRequest):
    """
    Trigger a proactive SMS sweep.

    Finds all users eligible for the given trigger type at the target
    hour in their local timezone, generates content via the TS
    orchestrator, and sends via SendBlue.

    Use trigger_type="all" to run all trigger types (for hourly cron).
    """
    # Verify internal auth
    secret = request.headers.get("x-internal-secret", "")
    if not INTERNAL_SMS_CHAT_SECRET or secret != INTERNAL_SMS_CHAT_SECRET:
        raise HTTPException(status_code=401, detail="Unauthorized")

    from background_tasks import run_proactive_sms_scheduler_job

    execution = await run_proactive_sms_scheduler_job(
        trigger_type=body.trigger_type,
        target_hour=body.target_hour,
    )
    logger.info("Proactive sweep occurrence result: %s", execution)
    if body.trigger_type == "all":
        return {
            "status": "ok",
            "occurrence_status": execution.status,
            "scheduled_for": execution.scheduled_for,
            "mode": "all",
            "sweeps": execution.result or [],
        }
    return {
        "status": "ok",
        "occurrence_status": execution.status,
        "scheduled_for": execution.scheduled_for,
        **(execution.result or {}),
    }

File: apps/dashboard/src/trigger/proactive-sms.ts
// src/trigger/proactive-sms.ts
/**
 * Proactive SMS Sweep — runs every hour.
 *
 * Calls the Python backend's proactive trigger endpoint with
 * trigger_type="all". The backend handles timezone-aware user
 * eligibility: for each trigger type (eod_recap at 9 PM,
 * morning_briefing at 8 AM, streak_alert at 8 PM, etc.), it
 * finds users whose local time matches the target hour,
 * generates content via the TS orchestrator, and sends via
 * SendBlue.
 *
 * One hourly cron covers all 5 trigger types.
 */

import { task, schedules } from "@trigger.dev/sdk/v3";
import { getTriggerBackendBaseUrl, triggerBackendFetch } from "@/lib/api/trigger-client";

async function runProactiveSweep() {
  const INTERNAL_SMS_CHAT_SECRET = process.env.INTERNAL_SMS_CHAT_SECRET;

  if (!INTERNAL_SMS_CHAT_SECRET) {
    throw new Error('INTERNAL_SMS_CHAT_SECRET environment variable is not set');
  }

  console.log('📬 Starting proactive SMS sweep...');
  console.log(`📊 Config: API=${getTriggerBackendBaseUrl()}`);

  const response = await triggerBackendFetch('/api/sms/proactive/trigger', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-internal-secret': INTERNAL_SMS_CHAT_SECRET,
    },
    body: JSON.stringify({
      trigger_type: 'all',
    }),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Proactive SMS sweep failed (${response.status}): ${errorText}`);
  }

  const result = await response.json();

  const totalSent = (result.sweeps || []).reduce(
    (sum: number, s: { sent?: number }) => sum + (s.sent || 0),
    0,
  );

  console.log('✅ Proactive SMS sweep completed');
  console.log(`📊 Results: ${totalSent} messages sent across ${(result.sweeps || []).length} trigger types`);

  return {
    success: true,
    timestamp: new Date().toISOString(),
    sweeps: result.sweeps,
    totalSent,
  };
}

/**
 * Manual trigger — can be invoked from the Trigger.dev dashboard.
 */
export const proactiveSmsTask = task({
  id: "proactive-sms-sweep",
  retry: {
    maxAttempts: 2,
    minTimeoutInMs: 2000,
    maxTimeoutInMs: 10000,
    factor: 2,
  },
  run: async () => {
    try {
      return await runProactiveSweep();
    } catch (error) {
      console.error('❌ Error running proactive SMS sweep:', error);
      throw error;
    }
  },
});

/**
 * Hourly schedule — runs at the top of every hour.
 * The Python backend determines which users are eligible based on
 * their local timezone and each trigger type's target hour.
 */
export const proactiveSmsSchedule = schedules.task({
  id: "proactive-sms-hourly",
  cron: "0 * * * *",
  run: async () => runProactiveSweep(),
});
